src/app/repositories/firebase/dataset_firebase_repository.py

```diff
@@ -35,44 +37,89 @@ def get_latest_dataset_with_survey_id_and_period_id(
 
         return dataset_metadata
 
-    def perform_new_dataset_transaction(
+    def perform_batched_dataset_write(
         self,
         dataset_id: str,
         dataset_metadata_without_id: DatasetMetadataWithoutId,
         unit_data_collection_with_metadata: list[UnitDataset],
         extracted_unit_data_identifiers: list[str],
-    ):
+    ) -> bool:
         """
-        Writes dataset metadata and unit data to firestore as a transaction, which is
-        rolled back if any of the operations fail.
-
+        Write dataset metadata and unit data to firestore in batches.
         Parameters:
-        dataset_id: id of the dataset
-        dataset_metadata_without_id: dataset metadata without a dataset id
-        unit_data_collection_with_metadata: collection of unit data associated to a dataset
-        extracted_unit_data_identifiers: identifiers associated with the unit data collection
+        dataset_id (str): The unique id of the dataset
+        dataset_metadata_without_id (DatasetMetadataWithoutId): The metadata of the dataset without its id
+        unit_data_collection_with_metadata (list[UnitDataset]): The collection of unit data associated with the new dataset
+        extracted_unit_data_identifiers (list[str]): List of identifiers ordered to match the identifier for each set of
+        unit data in the collection.
+
         """
 
-        # A stipulation of the @firestore.transactional decorator is the first parameter HAS
-        # to be 'transaction', but since we're using classes the first parameter is always
-        # 'self'. Encapsulating the transaction within this function circumvents the issue.
-        @firestore.transactional
-        def dataset_transaction(transaction: firestore.Transaction):
-            new_dataset_document = self.datasets_collection.document(dataset_id)
+        logger.info("Performing batch writes")
+        new_dataset_document = self.datasets_collection.document(dataset_id)
+        unit_data_collection_snapshot = new_dataset_document.collection("units")
+
+        try:
+            batch = self.client.batch()
+            batch.set(new_dataset_document, dataset_metadata_without_id, merge=True)
+            batch.commit()
+
+            batch_counter = 0
+            batch = self.client.batch()
+
+            for i in range(len(unit_data_collection_with_metadata)):
+                if batch_counter > 0 and batch_counter % self.BATCH_SIZE == 0:
+                    batch.commit()
+                    batch = self.client.batch()
```

Comment:
you are setting batch a second time, is this required?

Reply:
I think it is necessary to reset the batch after each commit based on this answer (although it is in Javascript, I just can't find a proper Python solution unfortunately)
Reference: https://stackoverflow.com/questions/66143646/run-multiple-firestore-batch-commit